### PR TITLE
Refactor superblock pack/unpack

### PR DESCRIPTION
**Pull request by denravonska**
**State: Ready**

Try to simplify superblock packing and unpacking to move away from the custom binary serialization code which was extremely slow.

TODO
====
- [x] Either remove boost dependency (replace with ntohs/htons) or bump the boost requirements.
 
Benchmark
========

Tested by measuring the time before and after running the pack/unpack in the unit tests. Times are in microseconds.

| Arch  | Before | After | Speedup |
|------|--------|------|-----------|
| i5 Pack      | 34003  | 1159  | 29.3x   |
| i5 Unpack    | 63120  | 403  | 156.6x   |
| Core2 Pack   |  63718 | 2212 | 28.8x |
| Core2 Unpack | 114080 | 621 | 183.7x |
| Pi 3 Pack         | ? | ? | ? |
| Pi 3 Unpack    | 566126 | 4416 | 128.2x |

63 milliseconds might not sound like that big of a deal, but we tally and unpack every 10 blocks so in a 500 block batch it adds up. Time saved on 500 v9 blocks:

- i5: 3.15s
- Core2: 5.7s
- RPi3: 28s

Help wanted
=========
The implementation still feels a bit messy so if anyone has any ideas on how to clean it up I'd be grateful.

I'd also like to test this before and after on a Raspbeery Pi to see what the speed increase looks like.

### CODE REVIEW

File: src/test/gridcoin_tests.cpp [hunk 1]
```diff
@@ -80,13 +83,27 @@ BOOST_AUTO_TEST_CASE(gridcoin_VerifyGetQuorumHash)
...
...6.25;e35f2e349c29296060d3f1ad52112d7d,20;e3609d2b817e65e79cdcba44c81e0538,0;e373665b11ba8869a310e5ffc8837b64,1.25;0,50;0,50;e3d2e26b764bda6c5e56338f09e5fb15,11.25;e3dbab06c42a46de0457314ca513c16e,22.5;e4027f17188b2a4ceda16cdbf1a8a965,0;e41cfa5a598151735ec6568fc94801ed,5;e41e9c4984d1640445e10bc3dd7da823,1.25;e474521b79aafeab8f5ce90a66df4aad,8.75;e47ec474e34e2256c426152b413101ac,1.25;e4b1ed8010a18456dbe158b6fdcdc0b5,1.25;0,50;e4d2d15b39f95f627094770b87e5b799,6.25;e4db42d47ed84bfbc5d492c6865bf6a9,0;e4eecfb50cae9c772fe597df7b2b8b8b,3.75;e4f4536c11db50b541309c4cc549d680,6.25;e503223b9d2ea770bef32e28c2fe3154,145;0,50;e52625f61028a2f4214a1010272c6754,8.75;e53253bd80d0c7c02ee6c8c44dc41e7b,45;e53ff8efab06a3bc0214dbf359006996,80;e546a0e0d5dc70b3819c59733ab7ff52,3.75;0,50;e59a6056bc2cbedf037e5d30fae0f293,6.25;e59d79b9fd799463f095ba14809dfde1,1.25;e5b4189e13aedee43992908678c6768a,100;e5c4f068970ea69d66293c52e7279117,50;e5cdc0c637cab4f382cbee77d9a9c2e7,55;e5d98ef2203fa1d1de8a6c618d4012e5,3.75;e5e0c31e740d02513dcc0f61f21195ca,2.5;e5e4fc4d91434914289aca9c26ebccd2,70;e602b53eba66086c5670bc7a0580598e,3.75;e60adf6eeb25da8aaf6ed2f460368189,35;e646fc78b3a208eafc93ed2a9c6fae6c,2.5;e64c6f0a50af958349a03
```
Comment (iFoggz) [inline src/test/gridcoin_tests.cpp]: UnPack* ;p

**Review by iFoggz — APPROVED** [review]
tACK; Approve

**Comment by denravonska** [comment]
This should also greatly help #768.

File: src/contract/superblock.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    // 1234567890123456MM
+    // MM = Magnitude stored as 2 bytes
+    // No delimiter between CPIDs, Step Rate = 18.
+    // CPID and magnitude are stored in big endian.
+    for (unsigned int x = 0; x < sBinary.length(); x += 18)
+    {
+        if(sBinary.length() - x < 18)
+            break;
+
+        const BinaryResearcher* researcher = reinterpret_cast<const BinaryResearcher*>(sBinary.data() + x);
```
Comment (tomasbrod) [inline src/contract/superblock.cpp:48]: This is endian unsafe and alignment unsafe.

**Review by tomasbrod — CHANGES_REQUESTED** [review]

File: src/contract/superblock.cpp (hunk 2 above)
Comment (denravonska) [reply src/contract/superblock.cpp:48]: It is endian safe (see calls to be16toh and htobe16) but not alignment safe. I'll see if we can add static asserts for it so we can deal with it as new platforms come along. Good catch.

**Comment by denravonska** [comment]
Added a static assert to ensure compiler error if BinaryResearcher::magnitude is offset.

**Pull request merged**